### PR TITLE
Implement location sharing for rooms

### PR DESCRIPTION
**Pull request by ctsk**
## Issue

This fixes the following issue(s):
- Resolves: #1518

## Screenshot

image

## Why this is useful for all students

It allows communicating locations on the campus even when the peer does not have the app themselves.

## Draft

This is currently a draft, as it merely allows to share rooms. When the viewed item is not a room, the button is not visible. I'd like to add the ability to share "buildings" (and other location types that "nav.tum.sexy" supports before merging.

### CODE REVIEW

**Comment by ctsk** [comment]
So this pretty much works now, however the button placement is not satisfactory yet.

**Comment by ctsk** [comment]
Update: Moved the button into the toolbar:

image

TODO: Find a proper android share icon

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt [hunk 1]
```diff
@@ -85,13 +85,29 @@ class NavigationDetailsFragment : BaseFragment<Unit>(
...
+        setupShareLocationButton(navigationDetails)
         setOpenInOtherAppBtnListener(navigationDetails)
 
         showNavigationDetailsProperties(navigationDetails)
 
         showAvailableMaps(navigationDetails)
     }
 
+    private fun setupShareLocationButton(navigationDetails: NavigationDetails) {
+        val url = "https://nav.tum.sexy/${navigationDetails.type}/${navigationDetails.id}";
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt:97]:
```suggestion
        val url = "https://nav.tum.sexy/${navigationDetails.type}/${navigationDetails.id}"
```

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt [hunk 2]
```diff
@@ -85,13 +85,29 @@ class NavigationDetailsFragment : BaseFragment<Unit>(
...
+
+        binding.toolbarNav.shareButton.setOnClickListener {
+            val sendIntent = Intent().apply {
+                action = Intent.ACTION_SEND
+                putExtra(Intent.EXTRA_TEXT, url)
+                type = "text/plain"
+            }
+
+            val shareIntent = Intent.createChooser(sendIntent, null);
+            startActivity(shareIntent);
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt:107]:
```suggestion
            val shareIntent = Intent.createChooser(sendIntent, null)
            startActivity(shareIntent)
```

**Review by CommanderStorm — COMMENTED** [review]
Formatting fixes

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt [hunk 3]
```diff
@@ -89,13 +89,29 @@ class NavigationDetailsFragment : BaseFragment<Unit>(
...
+        setupShareLocationButton(navigationDetails)
         setOpenInOtherAppBtnListener(navigationDetails)
 
         showNavigationDetailsProperties(navigationDetails)
 
         showAvailableMaps(navigationDetails)
     }
 
+    private fun setupShareLocationButton(navigationDetails: NavigationDetails) {
+        val url = "https://nav.tum.sexy/${navigationDetails.type}/${navigationDetails.id}"
```
Comment (CommanderStorm) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt:101]: This is sadly not 100% correct. Either we do this:
```suggestion
        val url = "https://nav.tum.sexy/view/${navigationDetails.id}"
```

Or steal this code:
https://github.com/TUM-Dev/NavigaTUM/blob/main/webclient/src/views/view/view-view.js#L24-L35

File: app/src/main/res/values-de/strings.xml [hunk 4]
```diff
@@ -781,6 +781,7 @@ Signatur: %5$s</string>
     <string name="loading_details">lädt</string>
     <string name="site_plans">Lagepläne</string>
     <string name="interactive_map">Interaktive Karte</string>
+    <string name="share_location">Ort teilen</string>
```
Comment (CommanderStorm) [inline app/src/main/res/values-de/strings.xml:784]: This translation is never used. 
Is this a mistake?

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
I have two changes,
Also: you did remove the share functionality in the last commit. Was that intentional

File: app/src/main/res/values-de/strings.xml (hunk 4 above)
Comment (ctsk) [reply app/src/main/res/values-de/strings.xml:784]: Yes, a mistake. It's a leftover from the first iteration where there was a button with text. (see screenshot in original pull request)

**Comment by ctsk** [comment]
> Also: you did remove the share functionality in the last commit. Was that intentional?

Did I? It's there on my end (when checking out the PR). I removed the large button and instead added an icon to the toolbar.

**Comment by CommanderStorm** [comment]
Ah, now I noticed it…

Honestly, not noticing this is probably something that most people will do.
I agree that two stacked buttons take up a lot of space.
Have you experimented with two horizontally stacked buttons with less text (f. ex. `share` and `open in`, maybe with an icon)?

**Comment by ctsk** [comment]
I've experimented with a bunch of layouts, and I think I like this one the most:

<details>
<summary> Screenshot </summary>
image
</details>

**Comment by ctsk** [comment]
Not sure how to say "Open in Maps" in German without it getting too long. Perhaps "Open with ..."?

**Comment by CommanderStorm** [comment]
> Unrelated: Is there a particular reason to use nav.tum.sexy and not nav.tum.de directly?

We got nav.tum.de like 2-3 Weeks ago, this is just an alias for now (nav.tum.sexy returns a permanent redirect, which the app follows)
Exchanging this string should probably be another PR 😉


> Not sure how to say "Open in Maps" in German without it getting too long. Perhaps "Öffnen mit ..."?

I would probably say `Maps öffnen` is the better option, but I am not a german expert (I barely got passing grades in the german Abitur...)

**Comment by CommanderStorm** [comment]
> I've experimented with a bunch of layouts, and I think I like this one the most

I think this is better, but I would like to experiment with a few other ideas (in terms of fewer margins, or maybe [something like this](https://medium.com/nerd-for-tech/how-to-add-extended-floating-action-button-in-android-android-studio-java-481cc9b3cdcb)). Could you push this version?

**Comment by ctsk** [comment]
<details>

<summary> Screenshot </summary>

image

</details>

Any ideas on how to avoid clashing with the site plans selection?

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt [hunk 5]
```diff
@@ -89,13 +89,29 @@ class NavigationDetailsFragment : BaseFragment<Unit>(
...
+        setupShareLocationButton(navigationDetails)
         setOpenInOtherAppBtnListener(navigationDetails)
 
         showNavigationDetailsProperties(navigationDetails)
 
         showAvailableMaps(navigationDetails)
     }
 
+    private fun setupShareLocationButton(navigationDetails: NavigationDetails) {
+        val url = "https://nav.tum.sexy/view/${navigationDetails.id}"
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt:101]: nav.tum.app pls

File: app/src/main/res/layout/toolbar_navigation.xml [hunk 6]
```diff
@@ -1,16 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (kordianbruck) [inline app/src/main/res/layout/toolbar_navigation.xml:1]: Is this an unrelated cleanup? I'd appreciate to put this into a separate PR, so its easier to roll back in case this actually still used somewhere.

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt (hunk 5 above)
Comment (CommanderStorm) [reply app/src/main/java/de/tum/in/tumcampusapp/component/tumui/roomfinder/NavigationDetailsFragment.kt:101]: Since we serve at `nav.tum.de`,
I think this would be better:
```suggestion
        val url = "https://nav.tum.de/view/${navigationDetails.id}"
```

**Review by CommanderStorm — APPROVED** [review]

**Comment by CommanderStorm** [comment]
I Agree.
Works as intended => merging

Thanks for this contribution
